demux/packet.c

```diff
@@ -242,3 +254,64 @@ int demux_packet_add_blockadditional(struct demux_packet *dp, uint64_t id,
         memcpy(sd + 8, data, size);
     return 0;
 }
+
+void demux_packet_pool_init(struct demux_packet_pool *pool, struct MPContext *parent)
+{
+    mp_mutex_init(&pool->lock);
+    pool->packets = NULL;
+    pool->parent = parent;
+}
+
+void demux_packet_pool_uninit(void *p)
+{
+    struct demux_packet_pool *pool = p;
+    if (!pool->parent->quit_fast)
+        demux_packet_pool_clear(pool);
+    mp_mutex_destroy(&pool->lock);
+}
+
+void demux_packet_pool_clear(struct demux_packet_pool *pool)
+{
+    mp_mutex_lock(&pool->lock);
+    struct demux_packet *dp = pool->packets;
+    while (dp) {
+        struct demux_packet *next = dp->next;
+        free_demux_packet(dp);
+        dp = next;
+    }
+    pool->packets = NULL;
+    mp_mutex_unlock(&pool->lock);
+}
+
+void demux_packet_pool_add(struct demux_packet_pool *pool,
+                           struct demux_packet *dp)
+{
+    if (!dp)
+        return;
+    dp->next = NULL;
+    demux_packet_pool_join(pool, dp, dp);
+}
+
+void demux_packet_pool_join(struct demux_packet_pool *pool,
+                            struct demux_packet *head, struct demux_packet *tail)
+{
+    if (!head)
+        return;
+    assert(tail);
+    assert(head != tail ? !!head->next : !head->next);
+
+    mp_mutex_lock(&pool->lock);
+    tail->next = pool->packets;
+    pool->packets = head;
+    mp_mutex_unlock(&pool->lock);
+}
+
+struct demux_packet *demux_packet_pool_pop(struct demux_packet_pool *pool)
+{
+    mp_mutex_lock(&pool->lock);
+    struct demux_packet *dp = pool->packets;
+    if (dp)
+        pool->packets = dp->next;
+    mp_mutex_unlock(&pool->lock);
```

Comment:
consider resetting `next` here for safety

Reply:
done